lib/event-libs/sdevent/sdevent.c

```diff
@@ -0,0 +1,400 @@
+    }
+
+    lws_pt_unlock(pt);
+    lws_context_unlock(pt->context);
+
+    lws_service_fd_tsi(context, &eventfd, wsi->tsi);
+
+    if (pt->destroy_self) {
+        lws_context_destroy(pt->context);
+        return -1;
+    }
+
+    // fire idle handler
+    struct sd_event_source *idletimer = pt_to_priv_sd(pt)->idletimer;
+    if (idletimer) {
+        sd_event_source_set_time(idletimer, (uint64_t) 0);
+        sd_event_source_set_enabled(idletimer, SD_EVENT_ON);
+    }
+
+    // allow further events
+    // Note: do not move the assignment up, lws_service_fd_tsi may invalidate it!
+    struct sd_event_source *watcher = wsi_to_priv_sd(wsi)->source;
+    if (watcher) {
+        sd_event_source_set_enabled(watcher, SD_EVENT_ONESHOT);
+    }
+
+    return 0;
+
+bail:
+    lws_pt_unlock(pt);
+    lws_context_unlock(pt->context);
+
+    return -1;
+}
+
+static void io_sd(struct lws *wsi, int flags) {
+
+    struct lws_context_per_thread *pt = &wsi->a.context->pt[(int)wsi->tsi];
+
+    // only manipulate if there is an event source and if the pt is still alive
+    if (!pt_to_priv_sd(pt)->io_loop || !wsi_to_priv_sd(wsi)->source || pt->is_destroyed) {
+        return;
+    }
+
+    // assert that the requested flags do not contain anything unexpected
+    if (!((flags & (LWS_EV_START | LWS_EV_STOP)) &&
+          (flags & (LWS_EV_READ | LWS_EV_WRITE)))) {
+        lwsl_err("%s: assert: flags %d", __func__, flags);
+        assert(0);
+    }
+
+    // we are overdoing a bit here, so it resembles the structure in libuv.c
+    if (flags & LWS_EV_START) {
+        if (flags & LWS_EV_WRITE)
+            wsi_to_priv_sd(wsi)->events |= EPOLLOUT;
+
+        if (flags & LWS_EV_READ)
+            wsi_to_priv_sd(wsi)->events |= EPOLLIN;
+
+        sd_event_source_set_io_events(wsi_to_priv_sd(wsi)->source, wsi_to_priv_sd(wsi)->events);
+        sd_event_source_set_enabled(wsi_to_priv_sd(wsi)->source, SD_EVENT_ONESHOT);
+    } else {
+        if (flags & LWS_EV_WRITE)
+            wsi_to_priv_sd(wsi)->events &= ~EPOLLOUT;
+
+        if (flags & LWS_EV_READ)
+            wsi_to_priv_sd(wsi)->events &= ~EPOLLIN;
+
+        sd_event_source_set_io_events(wsi_to_priv_sd(wsi)->source, wsi_to_priv_sd(wsi)->events);
+
+        if (!(wsi_to_priv_sd(wsi)->events & (EPOLLIN | EPOLLOUT)))
+            sd_event_source_set_enabled(wsi_to_priv_sd(wsi)->source, SD_EVENT_ONESHOT);
+        else
+            sd_event_source_set_enabled(wsi_to_priv_sd(wsi)->source, SD_EVENT_OFF);
+    }
+}
+
+static int init_vhost_listen_wsi_sd(struct lws *wsi) {
+    if (!wsi)
+        return 0;
+
+    struct lws_context_per_thread *pt = &wsi->a.context->pt[(int)wsi->tsi];
+
+    void *userdata = wsi;
+
+    sd_event_add_io(
+            pt_to_priv_sd(pt)->io_loop,
+            &wsi_to_priv_sd(wsi)->source,
+            wsi->desc.sockfd,
+            wsi_to_priv_sd(wsi)->events,
+            sock_accept_handler,
+            userdata
+    );
+
+    io_sd(wsi, LWS_EV_START | LWS_EV_READ);
```

Comment:
As I understand the value `wsi_to_priv_sd(wsi)->events` is only set in this function. But it is already used in the line before `sd_event_add_io`. Is this correct?

Reply:
Or is this intended to afterwards set the events with `sd_event_source_set_io_events`? But then we could also pass `0` or a specific variable like `const int initial_empty_events = 0` in the `sd_event_add_io` call to make this more clear.